## One Class SVM Classification

Classifies data as financially efficient or inefficient given training data of only efficient nonprofits. 

It defines a boundary based on the training data and classifies the data as positive(efficient) or negative(inefficient).


In [1]:
import pandas as pd


df1=pd.read_csv('team_out_1.csv')
df2=pd.read_csv('team_out_a2.csv')
df3=pd.read_csv('team_out_a3.csv')
df4=pd.read_csv('team_out_Yash.csv')
df5=pd.read_csv('team_out_Yash_part1.csv')


df=df1.append(df2)
df=df.append(df3)
df=df.append(df4)
df=df.append(df5)


df.dropna(inplace=True)
df.reset_index(inplace=True,drop=True)
df=df[df.Total_Expenses>0]
df

Filename          EIN  Program_Exp  Liabilities_To_Asset  \
0      201523229349300327.xml  510311790.0     0.989619              0.091802   
1      201543089349301829.xml  261460932.0     0.965378              0.000000   
2      201533179349306298.xml  270609504.0     0.942276              0.049206   
3      201533209349304768.xml  521548962.0     0.384526              0.000814   
4      201533179349307343.xml  731653383.0     0.620560              0.029487   
5      201533189349300608.xml  237324566.0     0.758902              0.736982   
6      201523069349301367.xml   43259150.0     0.736312              0.181998   
7      201533069349300963.xml  621273871.0     0.644822              0.234247   
8      201523099349300542.xml  541897455.0     0.845659              0.138531   
9      201533099349301033.xml  251869168.0     0.762357              0.000000   
10     201523169349304367.xml  376046335.0     0.745386              0.000000   
11     201533099349301803.xml   50454409.0     0.915261              0.000000   
12     201523069349300142.xml  272364809.0     0.920307              0.003606   
13     201543039349301204.xml  463812139.0     0.664088              0.648413   
14     201523089349301462.xml  561797737.0     0.704623              0.432671   
15     201533069349300788.xml  590638867.0     0.000000              0.000000   
16     201533079349300238.xml  273314429.0     1.000000              0.000000   
17     201523209349314257.xml  710801566.0     1.000000              0.000000   
18     201523209349311332.xml  752204594.0     0.896895              0.069715   
19     201533179349302173.xml  251374594.0     0.000000              0.313630   
20     201533179349307048.xml  450448773.0     0.991205              0.051348   
21     201533209349302633.xml  363789851.0     0.000000              0.232011   
22     201533099349301113.xml  411712804.0     0.477697              0.009887   
23     201523039349300127.xml  381557861.0     0.832159              0.063834   
24     201523079349301652.xml  273741627.0     0.945120              0.000000   
25     201533039349300813.xml  411495321.0     0.668161              0.040796   
26     201533139349300208.xml   60666277.0     0.000000              0.644904   
27     201533069349301413.xml   42616064.0     0.951218              0.338683   
28     201533079349300003.xml  581651220.0     0.910347              0.026832   
29     201523069349300957.xml  341496171.0     0.850617              0.068838   
...                       ...          ...          ...                   ...   
75744  201542929349301039.xml  205158717.0     0.490545              0.000000   
75745  201542929349301104.xml  440296401.0     0.000000              0.001255   
75746  201542929349301204.xml  131084330.0     0.000000              0.949950   
75747  201503109349302210.xml   30228267.0     0.726896              0.048491   
75748  201503109349302270.xml  741718905.0     0.000000              0.444349   
75749  201522549349300117.xml  150249365.0     0.000000              0.000000   
75750  201522549349300127.xml  942763918.0     0.000000              0.126873   
75751  201522549349300142.xml  453774366.0     0.000000              0.006878   
75752  201522549349300202.xml  860507921.0     0.751038              0.256169   
75753  201522549349300217.xml  236390816.0     0.000000              0.128085   
75754  201522549349300247.xml  940689854.0     1.000000              0.562191   
75755  201522549349300317.xml  830164620.0     0.000000              0.009078   
75756  201522549349300322.xml  274726130.0     0.991514              0.003792   
75757  201503139349301280.xml  810677279.0     0.564601              0.016695   
75758  201503139349301285.xml  232090256.0     0.767743              1.176003   
75759  201503139349301295.xml  202902396.0     0.009901              0.004141   
75760  201503139349301300.xml  222471909.0     0.940847              1.457766   
75761  201503139349301315.xml   50359008.0     0.634

In [634]:
df[df.Program_Exp>1]

Filename          EIN  Program_Exp  Liabilities_To_Asset  \
1545   201502579349301175.xml  200141248.0     1.142607              0.770354   
5627   201512949349301306.xml  542014609.0     1.013164              1.069741   
7960   201503359349300815.xml  273294817.0     1.010240              0.020753   
9800   201620279349300617.xml  550576920.0     1.006403              0.434611   
13111  201600949349300200.xml   10367116.0     1.003634              0.028478   
16056  201620629349300447.xml  542014609.0     1.013164              1.069741   
16692  201620419349300712.xml  311238139.0     1.053781              0.153538   
20638  201503169349304480.xml  363414823.0     9.982122              0.962854   

       Working_Capital  Surplus_Margin  Total_Expenses  
1545          0.553360       -5.089240       1276722.0  
5627         -0.011601       -0.276276       1574232.0  
7960          0.857745        0.424186       5355301.0  
9800          0.073114        0.013669       1822040.0  
13111         1.075774       -0.117847       1179223.0  
16056        -0.011601       -0.276276       1574232.0  
16692        34.644571       -6.766196       2024744.0  
20638         0.224504       -3.008420       2118845.0

In [2]:
small_df=df[df.Total_Expenses<1000000]

med_df=df[df.Total_Expenses>1000000]
med_df=med_df[df.Total_Expenses<10000000]

large_df=df[df.Total_Expenses<50000000]
large_df=large_df[df.Total_Expenses>10000000]

national_df=df[df.Total_Expenses>50000000]



/Users/araeyusvakil/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
/Users/araeyusvakil/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:7: UserWarning: Boolean Series key will be reindexed to match DataFrame index.


In [28]:
print(len(large_df))

pos_med=large_df[large_df.Program_Exp>.75]
# pos_med=pos_med[pos_med.Working_Capital>.01]
pos_med=pos_med[pos_med.Liabilities_To_Asset<1]
# pos_med=pos_med[pos_med.Surplus_Margin>.01]

lst_temp=list(pos_med['EIN'])
print("%:",len(lst_temp)/len(large_df))
print("NUMBER OF POSITIVE: ",len(lst_temp))

6054
%: 0.7381896266930955
NUMBER OF POSITIVE:  4469


In [3]:
print("NUMBER OF MED: ",len(med_df))

pos_med=med_df[med_df.Program_Exp>.75]
pos_med=pos_med[pos_med.Working_Capital>.01]
pos_med=pos_med[pos_med.Liabilities_To_Asset<1]
pos_med=pos_med[pos_med.Surplus_Margin>.01]
# 

lst_temp=list(pos_med['EIN'])
print("%:",len(lst_temp)/len(med_df))
print("NUMBER OF POSITIVE: ",len(lst_temp))

NUMBER OF MED:  23589
%: 0.3620755436856162
NUMBER OF POSITIVE:  8541


In [36]:
eff_nat_df=national_df[national_df.Program_Exp>.8]
# eff_nat_df=eff_nat_df[eff_nat_df.Working_Capital>.1]
eff_nat_df=eff_nat_df[eff_nat_df.Liabilities_To_Asset<1]
# eff_nat_df=eff_nat_df[eff_nat_df.Surplus_Margin>.1]

print(len(national_df))
lst_temp=list(eff_nat_df['EIN'])
print("%:",len(lst_temp)/len(national_df))
print(len(lst_temp))

2203
%: 0.7330912392192465
1615


In [18]:
print("NUMBER OF SMALL: ",len(small_df))

pos_small=small_df[small_df.Program_Exp>.5]
# pos_small=pos_small[pos_small.Working_Capital>.5]
pos_small=pos_small[pos_small.Liabilities_To_Asset<.5]
# pos_small=pos_small[pos_small.Surplus_Margin>.2]
lst_temp=list(pos_small['EIN'])

print("%:",len(lst_temp)/len(small_df))
print("NUMBER OF POSITIVE: ",len(lst_temp))


NUMBER OF SMALL:  43433
%: 0.6999516496673036
NUMBER OF POSITIVE:  30401


In [37]:
df=national_df #CHANGE THIS TO REQUIRED SIZE AND RUN REST OF THE CODE AS IS

In [38]:
df.reset_index(drop=True,inplace=True)
norm_df=df.copy()

norm_df=norm_df[['Program_Exp','Liabilities_To_Asset','Working_Capital','Surplus_Margin','Total_Expenses']]

from sklearn import preprocessing

x = norm_df.values #returns a numpy array
min_max_scaler = preprocessing.MinMaxScaler()
x_scaled = min_max_scaler.fit_transform(x)
norm_df = pd.DataFrame(x_scaled)
norm_df["Filename"]=df['Filename']

norm_df["EIN"]=df['EIN']
norm_df.columns=['Program_Exp','Liabilities_To_Asset','Working_Capital','Surplus_Margin','Total_Expenses','Filename','EIN']
norm_df.set_index('EIN',inplace=True)
norm_df

Program_Exp  Liabilities_To_Asset  Working_Capital  \
EIN                                                               
620497990.0     0.711697              0.285976         0.121912   
621674308.0     0.876123              0.285762         0.095081   
440579850.0     0.853395              0.287135         0.095332   
112625096.0     0.825931              0.286435         0.084335   
510216589.0     0.701615              0.287647         0.092581   
660258919.0     0.907194              0.286644         0.095035   
953766170.0     0.729660              0.286531         0.092281   
411404075.0     0.865371              0.285468         0.125450   
314390844.0     0.859851              0.285564         0.186793   
370635502.0     0.862693              0.286192         0.102220   
941156621.0     0.902687              0.286926         0.093936   
941080917.0     0.844000              0.286724         0.091394   
741287016.0     0.852722              0.285299         0.648553   
131628168.0     0.705233              0.285666         0.229296   
381359218.0     0.909901              0.285911         0.091507   
582155150.0     0.857106              0.286543         0.095117   
620636239.0     0.967490              0.285982         0.093331   
462134675.0     0.748392              0.285638         0.095488   
382776791.0     0.847499              0.286168         0.096974   
990109908.0     0.885965              0.286060         0.132521   
344428218.0     0.778457              0.286749         0.112824   
900532831.0     0.866060              0.288567         0.077672   
381359083.0     0.904910              0.285675         0.136528   
530240474.0     0.000000              0.286898         0.114032   
133783732.0     0.973119              0.287039         0.081489   
840772672.0     0.959278              0.285566         0.087844   
161533232.0     0.891779              0.287347         0.086287   
521209124.0     0.000000              0.288096         0.075535   
160968914.0     0.836507              0.287277         0.082881   
203238867.0     0.936855              0.287854         0.078382   
...                  ...                   ...              ...   
131921358.0     0.784374              0.285912         0.149928   
131740122.0     0.809940              0.286692         0.097239   
470662290.0     0.933724              0.286321         0.128206   
237315673.0     0.874673              0.285289         0.780224   
590594631.0     0.847908              0.285439         0.129852   
860098923.0     0.846156              0.286235         0.108920   
60646813.0      0.000000              0.287123         0.078338   
420698265.0     0.870026              0.286141         0.112828   
610523304.0     0.835605              0.286091         0.112672   
680480736.0     0.959980              0.287758         0.078976   
640723407.0     0.990448              0.287734         0.093831   
350867958.0     0.832801              0.286500         0.103267   
250965270.0     0.788830              0.287869         0.078629   
461056754.0     0.000000              0.287357         0.137552   
540853898.0     0.822107              0.285474         0.114367   
311551316.0     0.000000              0.288268         0.071456   
237385560.0     0.000000              0.285797         0.102417   
950743320.0     0.000000              0.285266         0.250977   
221750190.0     0.803019              0.286426         0.112178   
112458584.0     0.885544              0.287441         0.083550   
363992031.0     0.000000              0.287881         0.078338   
953797687.0     0.905737              0.285464         0.167191   
251469002.0     0.924633              0.285670         0.143026   
111672777.0     0.787968              0.287301         0.093674   
941606519.0     0.942571              0.286195         0.087211   
396084300.0     0.000000              0.285237         0.225581   
231857015.0     0.909018              0.287627         0.077

In [39]:
Y_class_df = pd.DataFrame()
X_class_df=norm_df.loc[lst_temp]
X_class_df['Efficiency'] = 1


Y_class_df['Efficiency'] = X_class_df['Efficiency'] 
X_class_df.drop('Efficiency', axis=1, inplace=True)


new_df=norm_df[['Program_Exp','Liabilities_To_Asset','Working_Capital','Surplus_Margin']]
X_class_df=X_class_df[['Program_Exp','Liabilities_To_Asset','Working_Capital','Surplus_Margin']]
# X_class_df=X_class_df.drop(X_class_df.index[2])   #OUTLIER REMOVER
X_class_df.reset_index(inplace=True,drop=True)

%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.font_manager
from sklearn import svm
from scipy import stats
from mpl_toolkits.mplot3d import Axes3D

from sklearn.decomposition import TruncatedSVD

svd = TruncatedSVD(n_components=2, n_iter=7)
reduced_df = svd.fit_transform(new_df)

#svd2 = TruncatedSVD(n_components=2, n_iter=7)
X_classtrain_df = svd.fit_transform(X_class_df)


'''outliers_fraction = 0.25'''
#colors = ['m', 'g', 'b']
clf = svm.OneClassSVM(nu=0.1, kernel="linear", gamma=.01,coef0=1.5)
clf.fit(X_classtrain_df)
#y_pred_test = clf.predict(normalized_df)
#print(y_pred_test)
Z1 = clf.predict(reduced_df)
res_matrix=Z1
print(res_matrix)
#print(reduced_df.shape)

if Z1.shape[0]%2==1:
    Z1=Z1[:-1]
temp_Z1=Z1
Z1 = Z1.reshape((-1,2))
print(Z1.shape)
xx1 = []
yy1= []
for i in reduced_df:
    xx1.append(i[0])
    yy1.append(i[1])
x1 = np.asarray(xx1)
y1 = np.asarray(yy1)
temp_y1=y1
temp_x1=x1
if len(x1)%2==1: #IS ODD:
    x1=x1[:-1]
if len(y1)%2==1:
    y1=y1[:-1]
x1 = x1.reshape((-1,2))
y1 = y1.reshape((-1,2))
print(x1.shape,y1.shape)

# plt.figure(0)
# plt.contourf(reduced_df[0:Z1.shape[0]], reduced_df[Z1.shape[0]:len(res_matrix)-1], Z1,cmap=plt.cm.coolwarm)
# plt.figure(1)
# plt.contourf(x1, y1, Z1)

# fig = plt.figure()
# ax = fig.add_subplot(111, projection='3d')
# ax.contourf(x1,y1,Z1)

# fig = plt.figure()
# ax = fig.add_subplot(111, projection='3d')
# ax.contour(reduced_df[0:Z1.shape[0]], reduced_df[Z1.shape[0]:len(res_matrix)-1], Z1,cmap=plt.cm.coolwarm)

# fig = plt.figure()
# ax = fig.add_subplot(111, projection='3d')
# ax.plot_trisurf(temp_x1,temp_y1, temp_Z1,cmap=plt.cm.coolwarm)
# # #plt.scatter(X_class_df.as_matrix()[:, 0], X_class_df.as_matrix()[:, 1], color='black')


[-1.  1.  1. ...,  1. -1.  1.]
(1101, 2)
(1101, 2) (1101, 2)


/Users/araeyusvakil/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()
/Users/araeyusvakil/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


In [40]:
new_df=norm_df.copy()
new_df.reset_index(inplace=True)

temp_list=[]

count=0
for i in range(len(res_matrix)):
    if res_matrix[i]==-1.0:
        temp_list.append(pd.DataFrame(df.loc[i]).transpose())
        count+=1
print(count)
print(count/len(norm_df))
ineff_nat_df=pd.concat(temp_list)
ineff_nat_df.reset_index(inplace=True,drop=True)
ineff_nat_df

623
0.28279618701770315


Filename          EIN Program_Exp Liabilities_To_Asset  \
0    201533179349305948.xml  6.20498e+08    0.718528             0.286403   
1    201523209349313822.xml  5.10217e+08    0.708349             0.912501   
2    201523209349311657.xml  9.53766e+08    0.736663             0.494316   
3    201503089349300305.xml  1.31628e+08    0.712002             0.170064   
4    201503039349300965.xml  4.62135e+08    0.755575             0.159723   
5    201533169349301178.xml  3.44428e+08    0.785929             0.575837   
6    201503099349301210.xml   5.3024e+08           0             0.631953   
7    201513089349300006.xml  5.21209e+08           0               1.0808   
8    201503099349300710.xml  3.11668e+08    0.720145             0.384355   
9    201513099349301401.xml  2.22325e+08    0.754058             0.384388   
10   201503079349301165.xml  5.30215e+08           0             0.211201   
11   201523209349309137.xml  5.91093e+08    0.725863             0.252772   
12   201533169349302458.xml   7.1025e+08    0.809637              0.35022   
13   201533169349301973.xml  9.30603e+08    0.754009             0.392576   
14   201533169349304798.xml   3.8149e+08    0.769776             0.611598   
15   201523149349301217.xml     5.36e+08           0               1.5614   
16   201532579349301178.xml  9.10567e+08    0.709724             0.483338   
17   201512879349300501.xml   1.3174e+08    0.762818             0.864261   
18   201532579349300048.xml  3.70661e+08    0.813327                    0   
19   201512999349300926.xml  4.11294e+08    0.717965             0.755306   
20   201522879349300742.xml  7.41031e+08    0.776701              0.41779   
21   201512869349301651.xml  9.51644e+08    0.754632            0.0450512   
22   201513029349301001.xml  3.40754e+08           0             0.175973   
23   201503019349300920.xml    5.302e+08           0             0.600912   
24   201513029349301666.xml  5.70752e+08    0.914972             -106.882   
25   201542999349301319.xml  5.91378e+08           0                    0   
26   201532879349300843.xml  5.22001e+08    0.661161              1.30181   
27   201532759349300858.xml  6.10582e+07           0             0.622961   
28   201502759349300960.xml  8.40399e+08    0.790303            0.0723272   
29   201502549349300340.xml   2.0233e+08     0.71991             0.458081   
..                      ...          ...         ...                  ...   
593  201533209349302483.xml  2.26106e+08    0.814969             0.429315   
594  201533209349302408.xml  7.50055e+08           0             0.797782   
595  201533209349302513.xml  3.40715e+08    0.820203             0.219401   
596  201543179349304149.xml  1.31628e+08           0             0.368225   
597  201523209349302112.xml  1.31624e+08    0.715179             0.372468   
598  201523209349302152.xml  6.11649e+08    0.813336             0.102526   
599  201513209349313236.xml  9.42736e+08    0.789357             0.741644   
600  201523179349306007.xml  2.51494e+08    0.781912             0.585004   
601  201523209349302462.xml  5.60928e+08    0.813337             0.053122   
602  201513209349311296.xml  9.11915e+08    0.795897              0.35823   
603  201523179349303937.xml   1.3174e+08     0.78755             0.989647   
604  201522939349300957.xml  8.50313e+08    0.686283               0.6969   
605  201512959349300826.xml  3.91028e+08    0.812847             0.520239   
606  201523209349300522.xml  5.21572e+08    0.791724             0.252599   
607  201523179349303117.xml  7.52679e+08    0.810856            -0.102514   
608  201503109349302150.xml  3.51968e+08    0.818064             0.467783   
609  201503109349301980.xml  2.50818e+08           0             0.912927   
610  201503169349302405.xml  1.31921e+08    0.791903             0.262411   
611  201522949349300642.xml   1.3174e+08    0.817714             0.554476   
612  201542939349300119.xml  6.06468e+07           0             0.716109   
613  2015131393

## Results

OneClassSVM determined these 623 National Sized Nonprofits as being financially inefficient, they are failing at one of the metrics, either their liabilities to assets ratio is too high or their program expenses ratio, working captial ratio or surplus margin is too low.